src/toil/batchSystems/kubernetes.py

```diff
@@ -130,11 +111,80 @@ def __init__(self, config, maxCores, maxMemory, maxDisk):
         # Set this to True to enable the experimental wait-for-job-update code
         self.enableWatching = True
 
-        # Required APIs needed from kubernetes
-        self.batchApi = kubernetes.client.BatchV1Api()
-        self.coreApi = kubernetes.client.CoreV1Api()
-    
         self.jobIds = set()
+        
+    def _api(self, kind, max_age_seconds = 5 * 60):
+        """
+        The Kubernetes module isn't clever enough to renew its credentials when
+        they are about to expire. See
+        https://github.com/kubernetes-client/python/issues/741.
+        
+        We work around this by making sure that every time we are about to talk
+        to Kubernetes, we have fresh credentials. And we do that by reloading
+        the config and replacing our Kubernetes API objects before we do any
+        Kubernetes things.
+        
+        TODO: We can still get in trouble if a single watch or listing loop
+        goes on longer than our credentials last, though.
+        
+        This method is the Right Way to get any Kubernetes API. You call it
+        with the API you want ('batch', 'core', or 'customObjects') and it
+        returns an API object with guaranteed fresh credentials.
+        
+        It also recognizes 'namespace' and returns our namespace as a string.
+        
+        max_age_seconds needs to be << your cluster's credential expiry time.
+        """
+        
+        now = datetime.datetime.now()
+        
+        if self.credential_time is None or (now - self.credential_time).total_seconds() > max_age_seconds:
```

Comment:
How do you feel about refreshing this a little before they expire, possibly a 10 second buffer?  Also... just a nit, but his credential refresh might be nice as its own function.

Reply:
We don't actually know how long the credentials last; I'm not sure we can query it and I think it is cluster-specific. Our cluster seems to let them last for an hour or so.

This PR just refreshes every 5 minutes, on the theory that that is Fast Enough.

Reply:
We could break this out into its own function but it would somehow have to report where it got the credentials from, which would be a bit weird.